src/imap_stream.rs

```diff
@@ -123,160 +77,350 @@ impl<R: Read + Write + Unpin> ImapStream<R> {
     pub fn as_mut(&mut self) -> &mut R {
         &mut self.inner
     }
+
+    /// End-Of-File return value.
+    ///
+    /// Return the appropriate EOF value for the stream depending on whether there is still
+    /// data in the buffer.  It is assumed that any remaining data in the buffer can not be
+    /// decoded.
+    fn stream_eof_value(&self) -> Option<io::Result<ResponseData>> {
+        match self.buffer.used() {
+            0 => None,
+            _ => Some(Err(io::Error::new(
+                io::ErrorKind::UnexpectedEof,
+                "bytes remaining in stream",
+            ))),
+        }
+    }
 }
 
 impl<R: Read + Write + Unpin> ImapStream<R> {
-    fn decode(
-        &mut self,
-        buf: Block<'static>,
-        start: usize,
-        end: usize,
-    ) -> io::Result<DecodeResult> {
-        log::trace!("decode: input: {:?}", std::str::from_utf8(&buf[start..end]));
-
-        let mut rest = None;
-        let mut used = 0;
-        let res = ResponseData::try_new(buf, |buf| {
-            match imap_proto::parse_response(&buf[start..end]) {
-                Ok((remaining, response)) => {
-                    // TODO: figure out if we can shrink to the minimum required size.
-                    self.decode_needs = 0;
-
-                    let mut buf = POOL.alloc(std::cmp::max(remaining.len(), INITIAL_CAPACITY));
-                    buf[..remaining.len()].copy_from_slice(remaining);
-                    used = remaining.len();
+    fn maybe_decode(&mut self) -> io::Result<Option<ResponseData>> {
+        if self.buffer.used() > self.decode_needs.unwrap_or(0) {
+            self.decode()
+        } else {
+            Ok(None)
+        }
+    }
 
-                    rest = Some(buf);
+    fn decode(&mut self) -> io::Result<Option<ResponseData>> {
+        let block: Block<'static> = self.buffer.take_block();
+        // Be aware, now self.buffer is invalid until block is returned or reset!
```

Comment:
Hmm, only self.buffer.block is now invalid.  self.buffer.used() is still valid and represents the used values in the block we have.  Perhaps there's better ways of writing this, e.g.
```
let used = self.buffer.used();
let block = self.buffer.take_block();
```
because that way we don't magically use something from self.buffer after we made it into a weird state.

Reply:
Or perhaps `.take_block()` should return a tuple of the block and offset/used_bytes_count.

Reply:
I wonder if it would be better to have sth like 


```rust
self.buffer.with_block(|block| {
  // code
});
```

to such that the api can enforce making it valid again

Reply:
That is indeed a much better idea, thanks!  I'll give that a go.

Reply:
I failed to make something work that is also more readable.  Problem is that the rental construction already uses a closure that mutably borrows self to set `self.decode_needs`.  To then also try and smuggle the reset block out of the rental construction closure becomes really hard.

So I'm fine with leaving the current version, it clearly has the critical section marked with two clear points where the state is back to good.  This makes is readable enough I think.